### PR TITLE
Add methods for structurally comparing syntax trees

### PR DESCRIPTION
**Pull request by bnbarham**
Adds a `findFirstDifference` API to `Syntax`, `XCTAssertSameStructure`
assertions, and a `SubtreeMatch` to compare subtrees.

### CODE REVIEW

**Comment by bnbarham** [comment]
The `SubtreeMatch` was originally a "contains", but I didn't love that we lose the nice tree-matching output in that case. So it's now either matching on the first node of that kind or the first node from some marker. That way we can still diff the entire subtree, assuming that a node with that type exists anyway. If it doesn't there's not much useful to print in any case 😅.

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

File: Tests/SwiftSyntaxTest/SyntaxComparisonTests.swift [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    let fd = SyntaxFactory.makeFunctionDecl(attributes: nil, modifiers: nil,
+                                            funcKeyword: keyword, identifier: identifier, genericParameterClause: nil,
+                                            signature: emptySignature, genericWhereClause: nil, body: funcBody)
+    if indent > 0 {
+      return fd.withLeadingTrivia(.spaces(indent))
+    }
+    return fd
+  }
+
+  private func makeBody(statementCount: Int = 0) -> CodeBlockSyntax {
```
Comment (ahoppen) [inline Tests/SwiftSyntaxTest/SyntaxComparisonTests.swift:167]: Suggested doc comment.

````suggestion
  /// Creates a `CodeBlockSyntax` that consists of `statementCount` integer literals with increasing values. I.e. `makeBody(statementCount: 2)` generates
  ```
  {
    0
    1
  }
  ```
  private func makeBody(statementCount: Int = 0) -> CodeBlockSyntax {
````

File: Tests/SwiftSyntaxTest/SyntaxComparisonTests.swift [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      XCTAssertEqual(diff.baseline.leadingTrivia, .spaces(2))
+      XCTAssertEqual(diff.node.leadingTrivia, [])
+    }
+
+    let actual = Syntax(makeFunc(identifier: SyntaxFactory.makeIdentifier("f")))
+    XCTAssertNil(actual.findFirstDifference(baseline: expected))
+    try expectations(actual.findFirstDifference(baseline: expected, includeTrivia: true))
+
+    let matcher = try SubtreeMatcher("struct A {func f() { }}")
+    XCTAssertNil(actual.findFirstDifference(baseline: expected))
```
Comment (ahoppen) [inline Tests/SwiftSyntaxTest/SyntaxComparisonTests.swift:73]: Should this be `matcher.findFirstDifference`? You already called 
```swift
XCTAssertNil(actual.findFirstDifference(baseline: expected)) 
```
above.

File: Sources/SwiftSyntax/SyntaxComparison.swift [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Utility APIs to compare two difference syntax trees for structual
+// equivalence.
+//
+//===----------------------------------------------------------------------===//
+
+public enum DifferenceReason {
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply Sources/SwiftSyntax/SyntaxComparison.swift:18]: Yeah, I can move it there. I had this before `_SwiftSyntaxTestSupport` was added but now that it has been... probably makes more sense there.

File: Sources/SwiftSyntax/SyntaxComparison.swift [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    if syntaxNodeType != baseline.syntaxNodeType {
+      return .nodeType
+    }
+
+    if isPresent != baseline.isPresent {
+      return .presence
+    }
+
+    if isToken {
+      if let token = `as`(TokenSyntax.self), let baselineToken = baseline.as(TokenSyntax.self) {
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply Sources/SwiftSyntax/SyntaxComparison.swift:127]: Ah sorry, you mentioned these in the previous one. I forgot to update them.

File: Sources/_SwiftSyntaxTestSupport/Syntax+Assertions.swift [hunk 5]
```diff
@@ -32,3 +33,207 @@ public func XCTAssertNext<Iterator: IteratorProtocol>(
...
+    var markers = [String: Int]()
+    var lastIndex = markedText.startIndex
+    for marker in findMarkedRanges(text: markedText) {
+      text += markedText[lastIndex ..< marker.range.lowerBound]
+      lastIndex = marker.range.upperBound
+      markers[String(marker.name)] = text.count
+    }
+    text += markedText[lastIndex ..< markedText.endIndex]
+
+    self.text = text.isEmpty ? markedText : text
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply Sources/_SwiftSyntaxTestSupport/Syntax+Assertions.swift:105]: Shouldn't happen, this was left over from a previous implementation. Thanks for catching!

File: Sources/_SwiftSyntaxTestSupport/Syntax+Assertions.swift [hunk 6]
```diff
@@ -32,3 +33,207 @@ public func XCTAssertNext<Iterator: IteratorProtocol>(
...
+  /// Same as `Syntax.findFirstDifference(baseline:includeTrivia:)`, but
+  /// matches against the first subtree from parsing `markedText` that is after
+  /// `afterMarker` with the root matching the root type of `baseline`.
+  public func findFirstDifference(afterMarker: String? = nil, baseline: Syntax, includeTrivia: Bool = false) throws -> TreeDifference? {
+    let afterMarker = afterMarker ?? markers.first!.key
+    guard let subtreeStart = markers[afterMarker] else {
+      throw SubtreeError.invalidMarker(name: afterMarker)
+    }
+
+    guard let possibleSubtree = SyntaxOffsetFinder.findFirstNode(in: actualTree, afterUTF8Offset: subtreeStart) else {
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply Sources/_SwiftSyntaxTestSupport/Syntax+Assertions.swift:119]: Yeah you're right, can just do this all in one (ie. find the first type after a given offset). Not sure why I split them now 🤔

File: Tests/SwiftSyntaxTest/SyntaxComparisonTests.swift [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+      }
+      """)
+    let funcDiff = try XCTUnwrap(matcher.findFirstDifference(afterMarker: "FUNC", baseline: expectedFunc))
+    XCTAssertEqual(funcDiff.reason, .additionalNode)
+
+    let bodyDiff = try XCTUnwrap(matcher.findFirstDifference(afterMarker: "BODY", baseline: expectedBody))
+    XCTAssertEqual(bodyDiff.reason, .additionalNode)
+  }
+
+  private func makeFunc(identifier: TokenSyntax, keyword: TokenSyntax = SyntaxFactory.makeFuncKeyword(),
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply Tests/SwiftSyntaxTest/SyntaxComparisonTests.swift:146]: Heh, I'm the opposite. I don't feel *super* strongly either way so if you really prefer these first I can do that.

I generally prefer "public" or "interesting" fields/methods first and then all the helpers/other junk after. That way when I'm reading the file for the first time I don't get bogged down in implementation details.

File: Sources/SwiftSyntax/Syntax.swift [hunk 8]
```diff
@@ -85,6 +85,76 @@ extension Syntax: Identifiable {
...
+    } else {
+      target.write(String(describing: syntaxNodeType))
+    }
+
+    let allChildren = children(viewMode: ._all)
+    if !allChildren.isEmpty {
+      target.write(" children=\(allChildren.count)")
+    }
+
+    if let converter {
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply Sources/SwiftSyntax/Syntax.swift:133]: We should probably set a minimum version if we want to enforce this.

File: Sources/SwiftSyntax/SyntaxComparison.swift [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    switch reason {
+    case .nodeType:
+      fallthrough
+    case .presence:
+      fallthrough
+    case .trivia:
+      fallthrough
+    case .tokenKind:
+      fallthrough
+    case .tokenText:
```
Comment (fwcd) [inline Sources/SwiftSyntax/SyntaxComparison.swift:56]: I think we can rewrite this to a slightly more-concise multi-case

```suggestion
    case .nodeType, .presence, .trivia, .tokenKind, .tokenText:
```

File: Sources/SwiftSyntax/Trivia.swift.gyb [hunk 10]
```diff
@@ -134,6 +134,15 @@ public struct Trivia {
 % end
 }
 
+extension Trivia: CustomDebugStringConvertible {
+   public var debugDescription: String {
+     if count == 1, let first {
+       return first.debugDescription
+     }
+     return "[" + map({ $0.debugDescription }).joined(separator: ", ") + "]"
```
Comment (fwcd) [inline Sources/SwiftSyntax/Trivia.swift.gyb:142]:
```suggestion
     return "[" + map(\.debugDescription).joined(separator: ", ") + "]"
```

File: Sources/SwiftSyntax/gyb_generated/Trivia.swift [hunk 11]
```diff
@@ -260,6 +260,15 @@ public struct Trivia {
   }
 }
 
+extension Trivia: CustomDebugStringConvertible {
+   public var debugDescription: String {
+     if count == 1, let first {
+       return first.debugDescription
+     }
+     return "[" + map({ $0.debugDescription }).joined(separator: ", ") + "]"
```
Comment (fwcd) [inline Sources/SwiftSyntax/gyb_generated/Trivia.swift:268]:
```suggestion
     return "[" + map(\.debugDescription).joined(separator: ", ") + "]"
```

File: Sources/SwiftSyntax/Syntax.swift [hunk 12]
```diff
@@ -85,6 +85,76 @@ extension Syntax: Identifiable {
   public typealias ID = SyntaxIdentifier
 }
 
+extension Syntax: CustomDebugStringConvertible {
+  public var debugDescription: String {
+    return debugDescription()
```
Comment (fwcd) [inline Sources/SwiftSyntax/Syntax.swift:90]:
```suggestion
    debugDescription()
```

**Review by fwcd — COMMENTED** [review]
Looks exciting, I have added a few (minor) style suggestions below.

File: Sources/SwiftSyntax/SyntaxComparison.swift (hunk 9 above)
Comment (bnbarham) [reply Sources/SwiftSyntax/SyntaxComparison.swift:56]: Oof yeah, thanks. I clearly have too much C-language-style in me 😆.

File: Sources/SwiftSyntax/gyb_generated/Trivia.swift (hunk 11 above)
Comment (bnbarham) [reply Sources/SwiftSyntax/gyb_generated/Trivia.swift:268]: Gyb-generated so will be fixed by the above one 🙇

Comment (fwcd) [reply Sources/SwiftSyntax/gyb_generated/Trivia.swift:268]: Ah, I missed that this was a generated file 😄

**Comment by bnbarham** [comment]
This currently includes https://github.com/apple/swift-syntax/pull/541 but I'll remove it after that one is in.

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
The `CustomReflectableTests` currently fail because of the new `debugDescription` in `SyntaxProtocol` and `Trivia`. I was going to update them, but ... do we actually still want all the custom reflectable bits? If we do still want it, it might be worth cleaning it up a bit more. Eg. not including nil children, not including `pieces` in `Trivia` (since that's now just part of the debug description for `Trivia`), collapsing `TokenSyntax`, add `missing` for `SyntaxProtocol` etc.

I'd still like to keep the expanded `debugDescription` either way since it's nice to only include trivia if it's important, mark specific nodes, and add locations in.

**Comment by bnbarham** [comment]
I've just changed the the test for now and will remove in another PR.

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

File: Sources/_SwiftSyntaxTestSupport/Syntax+Assertions.swift [hunk 13]
```diff
@@ -32,3 +33,192 @@ public func XCTAssertNext<Iterator: IteratorProtocol>(
...
+  fileprivate override func visitAny(_ node: Syntax) -> SyntaxVisitorContinueKind {
+    if found != nil {
+      return .skipChildren
+    }
+
+    if node.positionAfterSkippingLeadingTrivia.utf8Offset >= offset && node.syntaxNodeType == type {
+      found = node
+      return .skipChildren
+    }
+
```
Comment (ahoppen) [inline Sources/_SwiftSyntaxTestSupport/Syntax+Assertions.swift:215]: A little performance optimization suggestion (not sure if it makes much of a difference in practice): We don’t need to walk the children if the node’s end location is before `offset`

```swift
if node.endPosition.utf8Offset < offset {
  return .skipChildren
}
```

**Review by ahoppen — APPROVED** [review]
One minor suggestion that I don’t have too strong opinions about whether we should take it in. Otherwise, let’s 🚢it

File: Sources/_SwiftSyntaxTestSupport/Syntax+Assertions.swift (hunk 13 above)
Comment (bnbarham) [reply Sources/_SwiftSyntaxTestSupport/Syntax+Assertions.swift:215]: Ah yeah, good point. Done.

**Comment by bnbarham** [comment]
@swift-ci please test

File: Tests/SwiftSyntaxTest/SyntaxComparisonTests.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (CodaFi) [reply Tests/SwiftSyntaxTest/SyntaxComparisonTests.swift:151]: This repo still has clients that expect to be able to build us with Xcode 13, so for the time being we cannot adopt new features like this.

**Pull request merged**